row_group.go

```diff
@@ -303,112 +329,114 @@ func (r *rowGroupRows) Close() error {
 		r.done = nil
 	}
 
-	for i := range r.columns {
-		if err := r.columns[i].close(); err != nil {
+	for i := range r.readers {
+		if err := r.readers[i].Close(); err != nil {
 			lastErr = err
 		}
 	}
 
+	r.clear()
 	r.inited = true
 	r.closed = true
 	return lastErr
 }
 
-func (r *rowGroupRows) Schema() *Schema {
-	return r.rowGroup.Schema()
-}
-
 func (r *rowGroupRows) SeekToRow(rowIndex int64) error {
+	var lastErr error
+
 	if r.closed {
 		return io.ErrClosedPipe
 	}
 
-	for i := range r.columns {
-		if err := r.columns[i].seekToRow(rowIndex); err != nil {
-			return err
+	if !r.inited {
+		r.init()
+	}
+
+	for i := range r.readers {
+		if err := r.readers[i].SeekToRow(rowIndex); err != nil {
+			lastErr = err
 		}
 	}
 
-	r.seek = rowIndex
-	return nil
+	r.clear()
+	return lastErr
 }
 
 func (r *rowGroupRows) ReadRows(rows []Row) (int, error) {
+	if r.closed {
+		return 0, io.EOF
+	}
+
 	if !r.inited {
 		r.init()
-		if r.seek > 0 {
-			if err := r.SeekToRow(r.seek); err != nil {
-				return 0, err
+	}
+
+	// Limit the number of rows that we read to the smallest number of rows
+	// remaining in the current page of each column. This is necessary because
+	// the pointers exposed to the returned rows need to remain valid until the
+	// next call to ReadRows, SeekToRow, Reset, or Close. If we release one of
+	// the columns' page, the rows that were already read during the ReadRows
+	// call would be invalidated, and might reference memory locations that have
+	// been reused due to pooling of page buffers.
+	//
+	// The application is expected to handle the case where ReadRows returns
+	// less rows that requested and no error, by looking at the first returned
+	// value from ReadRows, which is the number of rows that were read.
```

Comment:
I like these comments, they're very clear!  Should we add/move this second paragraph to the `func (r *rowGroupRows) ReadRows` godoc though?

Also, minor grammar fix: "less rows that requested" -> "less rows *than* requested".